Rakefile

```diff
@@ -1,24 +1,70 @@
+
+# include path for the legacy and main protobuf compilers respectivicey
+toolchain_include_dir=File.join(toolchain_dir, "/include/proto")
+toolchain_legacy_include_dir=File.join(toolchain_dir, "/include/proto_legacy")
 
 protoc_legacy_version="3.5.1"
-protoc_legacy_binary="/tmp/protoc#{protoc_legacy_version}"
+protoc_legacy_binary=File.join(toolchain_bin_dir, "/protoc" + protoc_legacy_version)
 
 protoc_version="21.12"
-protoc_binary="/tmp/protoc#{protoc_version}"
+protoc_binary=File.join(toolchain_bin_dir, "/protoc" + protoc_version)
 
-gogo_dir="/tmp/gogo"
-protoc_gen_go_dir="/tmp/protoc-gen-go"
+gogo_tag = "v1.3.2"
+gogo_dir=File.join(toolchain_dir,  "/gogo")
+gogo_include = "#{toolchain_dir}/gogo/src:#{toolchain_dir}/gogo/src/github.com/gogo/protobuf/protobuf"
+gogo_bin = File.join(toolchain_bin_dir, "/gogo-bin-#{gogo_tag}")
 protoc_jsonschema_version="73d5723"
 
+### toolchains is meant to store a cache of all binary dependencies needed to build the agent-payload.
+### this rakefile will download those dependencies on the fly if needed.
+### the structure of toolchains is as follows:
+### /toolchains
+###     /toolchains/bin -- contains any binaries used for building
+###         /toolchains/bin/protoc21.12 -- the new protobuf compiler
+###         /toolchains/bin/protoc3.5.1 -- the legacy protobuf compiler
+###         /toolchains/bin/protoc-gen-go -- the go code generator for protoc
+###         /toolchains/bin/protoc-gen-XXX -- other protobuf generators (vtproto, java, etc.)
+###     /toolchains/include -- contains any protobuf library files for common types (like proto.Duration)
+###         /toolchains/includes/proto -- protobuf libraries bundled with the new protobuf compiler
+###         /toolchains/includes/proto_legacy -- protobuf libraries bundled with the legacy protobuf compiler
+### =>  /toolchains/gogo -- temp directory used to build the gogo_faster generator
 
 namespace :codegen do
+  task :clean do
+      sh "rm -rf #{gogo_dir}"
+      sh "rm -rf #{toolchain_dir}"
+  end
 
-  task :install_protoc do
-    if `bash -c "protoc --version"` != "libprotoc ${protoc_legacy_version}"
+  task :setup_gogo => ['install_protoc_all'] do
+    if ! Dir.exist?(gogo_dir) then
+      directory "#{gogo_dir}/src/github.com/gogo"
+      sh "git clone https://github.com/gogo/protobuf.git #{gogo_dir}/src/github.com/gogo/protobuf"
+    else
+      puts "gogo already cloned into #{gogo_dir}"
+    end
+
+    Dir.chdir("#{gogo_dir}/src/github.com/gogo/protobuf") do
+      sh "git checkout #{gogo_tag}"
+      sh "PATH=$PATH:/tmp  GOBIN=#{toolchain_bin_dir} GOPATH=#{gogo_dir} make clean install"
+    end
+  end
+
+
+  task 'install_protoc_all' => ['install_protoc', 'install_protoc_legacy']
+
+  task :install_protoc_legacy  do
+    if `bash -c "#{protoc_legacy_binary} --version"` != "libprotoc ${protoc_version}"
       sh <<-EOF
         /bin/bash <<BASH
+
+        set -euo pipefail
         if [ ! -f #{protoc_legacy_binary} ] ; then
```

Comment:
Bit of odd logic here. Aren't we checking the existence of `libprotoc ${protoc_version` in the `if` check above? I supose it could be the handling of special case where someone already has libprotoc installed but in the old location?

Reply:
It is. This is existing code that looks new because of git diff.

Ideally this would all be done with ruby primitives, but I didn't want to add even more to this PR.